archive/backup.go

```diff
@@ -0,0 +1,145 @@
+package archive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/0xPolygon/polygon-sdk/helper/common"
+	"github.com/0xPolygon/polygon-sdk/server/proto"
+	"github.com/0xPolygon/polygon-sdk/types"
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+func CreateBackup(conn *grpc.ClientConn, logger hclog.Logger, targetFrom uint64, targetTo *uint64, outPath string) (uint64, uint64, error) {
+	// always create new file, throw error if the file exists
+	fs, err := os.OpenFile(outPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	signalCh := common.GetTerminationSignalCh()
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+	go func() {
+		<-signalCh
+		logger.Info("Caught termination signal, shutting down...")
+		cancelFn()
+	}()
+
+	clt := proto.NewSystemClient(conn)
+
+	to, toHash, err := determineTo(ctx, clt, targetTo)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	stream, err := clt.Export(ctx, &proto.ExportRequest{
+		From: targetFrom,
+		To:   to,
+	})
+	if err != nil {
+		return 0, 0, err
+	}
+
+	writeMetadata(fs, logger, to, toHash)
+	resFrom, resTo, err := processExportStream(stream, logger, fs, targetFrom, to)
+
+	fs.Close()
+	if err != nil {
+		os.Remove(outPath)
```

Comment:
unhandled error? or you don't care?

Reply:
Hello, Thank you for your kind suggestion. That was my fault and I should have cared about this as you said. I've added an error handling at this line at this commit d4e7bc76ce1ae4ae5c5f7c616c2fe64340109a51.